[DOCS-15222] Document Java agentless Feature Flags delivery (#38518)
DataDog/documentation: content/en/feature_flags/server/java.md

Reviewer: ```suggestion
```
This PR removes the only reference to this link, so we can remove it here 🙂

- link: "/feature_flags/concepts/flag_graphs/"
  tag: "Concept"
  text: "Feature Flag Graphs"
- link: "/feature_flags/concepts/configuration_sources/"
  tag: "Concept"
  text: "Server SDK Configuration Sources"
---

## Overview

This page describes how to instrument a Java application with the Datadog Feature Flags SDK. Starting in version 1.65.0, the SDK loads flag configuration directly from the Datadog-managed CDN by default.

The Java SDK integrates with the Datadog Java tracer (`dd-trace-java`) and implements the [OpenFeature](https://openfeature.dev/) standard.

> <div class="alert alert-warning">In agentless mode, Java 1.65.0 supports configuration delivery and local flag evaluation only. It does not support evaluation metrics, exposure logging, or experimentation use cases.</div>

Reviewer: ```suggestion
<div class="alert alert-warning">Starting in version 1.65.0, in agentless mode, Java supports configuration delivery and local flag evaluation only. It does not support evaluation metrics, exposure logging, or experimentation use cases.</div>
```
Borrowing the "Starting in..." language from earlier on the page, so it doesn't sound like this only applies to just the one version.

> <div class="alert alert-warning">Java 1.65.0 supports agentless configuration delivery and local flag evaluation only. It does not support evaluation metrics, exposure logging, or experimentation use cases in agentless mode.</div>

Author: Applied: led with the agentless-mode qualifier in d75bee81d1.

## Compatibility requirements

For the default agentless setup, you need:

- **Java 11 or higher**
- **Datadog Java SDK** (`dd-java-agent`, added with `-javaagent`): Version **1.65.0** or later
- **Datadog OpenFeature provider** (`com.datadoghq:dd-openfeature`, added as a build dependency): Version **1.65.0** or later
- **OpenFeature SDK**: Version **1.20.1** or later
- A Datadog [API key][7]

> - Your Datadog site

Reviewer: ```suggestion
- A Datadog [**API key**][7]
- Your Datadog [**site**][14]
```
For consistency, adding bolding (which I think is kind of unnecessary for this list, but that's neither here nor there) and a link in case customers want some more info on sites (I also suggested the link reference at the bottom of this file)

The Datadog provider implements the [OpenFeature](https://openfeature.dev/) standard. It uses `dd-java-agent` for configuration delivery. Agentless delivery removes the external Datadog Agent requirement, but `dd-java-agent` must still load in the JVM.

<div class="alert alert-warning">Starting in version 1.65.0, agentless mode changes only flag configuration. Java still requires a supported Datadog Agent or serverless telemetry path to export evaluation metrics or exposure events. Without such a path, only configuration delivery and local flag evaluation work.</div>

## Compatibility requirements

For the default agentless setup, you need:

- **Java 11 or higher**
- **Datadog Java agent** (`dd-java-agent`, loaded with `-javaagent`): Version **1.65.0** or later
- **Datadog OpenFeature provider** (`com.datadoghq:dd-openfeature`, added as a build dependency): Version **1.65.0** or later
- **OpenFeature SDK**: Version **1.20.1** or later
- A Datadog [**API key**][7]
- Your Datadog [**site**][14]

> Use the same version of `dd-java-agent` and `dd-openfeature`. Agentless delivery does not require a separate Datadog Agent service.

Reviewer 2: Should this be use a minimum version? What if `dd-java-agent` is upgraded but we haven't made any changes to `dd-openfeature`?

Author: I mention the minimum version 1.65.0 (next minor) in a few places; they are published w/ the same tag.

* https://mvnrepository.com/artifact/com.datadoghq/dd-java-agent
* https://mvnrepository.com/artifact/com.datadoghq/dd-openfeature

Reviewer 2: >Use the same version of `dd-java-agent` and `dd-openfeature`

I was wondering if this may make people think `dd-java-agent` and `dd-openfeature` must always be the same version as each other.

Author: I see - for this release specifically yes agent just needs to be bumped to 1.6.5. 

In the serverless docs (https://docs.datadoghq.com/serverless/guide/datadog_forwarder_java/), they encourage people to always download latest Java agent, or rather don't really mention a pinned version.

Overall it is a positive thing for us if people DO keep the versions in sync, for example in the next release we'll add exposure forwarding through the Java agent and I want them to upgrade to it.

The Gradle and Maven installation examples pin specific versions of `dd-openfeature` and the OpenFeature SDK. See [Compatibility requirements](#compatibility-requirements) for the minimum supported versions.

> Flag evaluation metrics require a supported Agent-backed configuration. See [Set Up Server-Side Flag Evaluation Metrics][8].

Author: Kept the generic Agent-backed wording. Evaluation metrics require an Agent telemetry path, not necessarily Agent Remote Configuration.

2. The flag targets the **correct environment** (`DD_ENV`). Flags do not target specific services—they apply to all services within the enabled environment.
3. Your `DD_ENV` value appears in [{{< ui >}}Feature Flag Environments{{< /ui >}}][5]. If it is absent, the environment has not received any flag traffic yet.

### 2. Verify the configuration source

#### Agentless

1. Confirm that `dd-java-agent` and `dd-openfeature` are version 1.65.0 or later.

> 2. Confirm that `DD_FEATURE_FLAGS_ENABLED` is not set to `false`.

Reviewer: ```suggestion
2. Confirm that `DD_FEATURE_FLAGS_ENABLED` is set to `true`.
```
Can we say this, or is it possible for it to be null? Just thought that this would be a little easier to parse

3. Confirm that `DD_FEATURE_FLAGS_CONFIGURATION_SOURCE=agentless` is set, or that the source and legacy provider settings are not set.
4. Confirm that application code initializes the Datadog OpenFeature provider.
5. Confirm that `DD_API_KEY`, `DD_SITE`, and `DD_ENV` are configured in the application process.
6. Confirm that the application can make outbound HTTPS requests to Datadog.
7. Enable `DD_TRACE_DEBUG=true` and check for authentication, timeout, or malformed-payload messages from the Feature Flags agentless endpoint.

#### Agent Remote Configuration

1. Confirm that `DD_FEATURE_FLAGS_CONFIGURATION_SOURCE=remote_config` is set. During the migration window, `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true` also selects Remote Configuration when no source is set.

> 2. Confirm that `DD_FEATURE_FLAGS_ENABLED` is not set to `false`.

Reviewer: ```suggestion
2. Confirm that `DD_FEATURE_FLAGS_ENABLED` is set to `true`.
```
Same question as above 🙂

3. Confirm that Agent 7.55 or later is running and reachable. See [APM Connection Errors][2].
4. Confirm that Remote Configuration is enabled on the Agent. If it is disabled, set `remote_configuration.enabled: true` in `datadog.yaml` or `DD_REMOTE_CONFIGURATION_ENABLED=true`. See [Remote Configuration][1].
5. Confirm that `DD_API_KEY` is valid on the Agent and belongs to the target organization.
6. Confirm that `DD_SITE` is set correctly on the Agent. See [Agent Site Issues][3].
7. Run `datadog-agent status` and review the Remote Configuration section. See [Agent Commands][6].

### 3. SDK: Verify Java SDK state

#### Enable debug logging

All Feature Flags startup messages are emitted at DEBUG level. Set `DD_TRACE_DEBUG=true`. For the default agentless source, check this startup sequence:

```
[dd.trace] Feature Flagging system starting
[dd.trace] Feature Flagging system awaiting application provider activation
```

After the application initializes the provider, confirm that the system starts:

```
[dd.trace] Feature Flagging system started
```

> With `remote_config`, the system starts during tracer initialization and does not wait for provider activation. If the applicable messages are absent, confirm that `DD_FEATURE_FLAGS_ENABLED` is not `false`. Then verify the selected source.

Reviewer: ```suggestion
With `remote_config`, the system starts during tracer initialization and does not wait for provider activation. If the applicable messages are absent, confirm that `DD_FEATURE_FLAGS_ENABLED` is `true`. Then verify the selected source.
```
Same as above!

2. The flag targets the **correct environment** (`DD_ENV`). Flags do not target specific services—they apply to all services within the enabled environment.
3. Your `DD_ENV` value appears in [{{< ui >}}Feature Flag Environments{{< /ui >}}][5]. If it is absent, the environment has not received any flag traffic yet.

### 2. Verify the configuration source

#### Agentless

1. Confirm that `dd-openfeature` and `dd-java-agent` are version 1.65.0 or later. Use the same version for both components.
2. Confirm that the JVM loads `dd-java-agent` with `-javaagent`, either in the Java command or through `JAVA_TOOL_OPTIONS`.
3. Confirm that `DD_FEATURE_FLAGS_ENABLED` is unset or set to `true`.
4. Confirm that `DD_FEATURE_FLAGS_CONFIGURATION_SOURCE=agentless` is set, or that the source and legacy provider settings are not set.
5. Confirm that application code initializes the Datadog OpenFeature provider.
6. Confirm that `DD_API_KEY`, `DD_SITE`, and `DD_ENV` are configured in the application process.
7. Confirm that the application can make outbound HTTPS requests to Datadog.

> 8. Enable `DD_TRACE_DEBUG=true` and check for authentication, timeout, or malformed-payload messages from the Feature Flags agentless endpoint.

Reviewer 2: Should we add something like during the migration window confirm `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED` is no longer set

Author: The client will log when there is bad configuration: https://github.com/DataDog/dd-trace-java/blob/9a23583a981ef90d386299f9be3d54b6eb1dec3d/products/feature-flagging/feature-flagging-agent/src/main/java/com/datadog/featureflag/FeatureFlaggingSystem.java#L74

`TYPE_MISMATCH` is returned when the evaluation method does not match the flag's configured type. Use the correct method for each flag type: `getBooleanValue()`, `getStringValue()`, `getIntegerValue()`, `getDoubleValue()`.

### 4. Flagging platform: Verify data appears in Datadog

> <div class="alert alert-warning">Java 1.65.0 does not send exposure events or the <code>feature_flag.evaluations</code> metric without an Agent. Their absence in a no-Agent deployment is expected and does not indicate that configuration loading or local evaluation failed.</div>

Reviewer: ```suggestion
<div class="alert alert-warning">Starting in version 1.65.0, Java does not send exposure events or the <code>feature_flag.evaluations</code> metric without an Agent. Their absence in a no-Agent deployment is expected and does not indicate that configuration loading or local evaluation failed.</div>
```
Same as above, trying to avoid the misconception that this only applies to this one version

> <div class="alert alert-warning">Java 1.65.0 does not provide agentless delivery for exposure events or aggregate <code>flagevaluation</code> events. Their absence in a no-Agent deployment is expected and does not indicate that configuration loading or local evaluation failed.</div>

Author: Applied: used the canonical feature_flag.evaluations metric name and clarified the no-Agent behavior in d75bee81d1.

#### Flag evaluation metrics

Flag evaluation counts appear in Datadog as a `feature_flag.evaluations` counter metric tagged with the flag key, result variant, and evaluation reason. See <a href="/feature_flags/guide/server_flag_evaluation_metrics/">Set Up Server-Side Flag Evaluation Metrics</a> for the full setup guide and troubleshooting steps.

#### Experiment exposures

> When the selected telemetry path supports exposures, exposures appear only for flags associated with an experiment. Standard feature flags do not generate exposure events. If exposures are missing:

Reviewer: ```suggestion
When the selected configuration path supports exposures, exposures appear only for flags associated with an experiment. Standard feature flags do not generate exposure events. If exposures are missing:
```
Claude suggested using this instead of "telemetry path," which doesn't appear anywhere else and that people might misunderstand. I'm okay with either, but thought I'd flag it just in case.

[6]: /agent/configuration/agent-commands/
[7]: /account_management/api-app-keys/#api-keys
[8]: /feature_flags/guide/server_flag_evaluation_metrics/
[9]: /feature_flags/concepts/configuration_sources/
[10]: /feature_flags/concepts/configuration_sources/#use-a-custom-agentless-endpoint
[11]: /feature_flags/concepts/configuration_sources/#migrate-an-existing-remote-configuration-setup
[12]: /feature_flags/concepts/configuration_sources/#configure-agentless-delivery

> [13]: /feature_flags/concepts/configuration_sources/#use-agent-remote-configuration

Reviewer: ```suggestion
[13]: /feature_flags/concepts/configuration_sources/#use-agent-remote-configuration
[14]: /getting_started/site/
```
For the above link I added
